File: mac/Marquee/API/ServerEvents.swift
import Foundation
import Observation

/// Change counters screens key their `.task(id:)` reloads off, the server-era
/// replacement for the old engine's `AppEvents`. Every successful mutation
/// through `MarqueeAPI` bumps the areas it touches; `LiveUpdates` bumps
/// `notifications` / `requests` when polling sees the counts move on the
/// server (a download finished, a member requested something from the website).
///
///     .task(id: model.events.revision(of: [.library, .requests])) { await load() }
@MainActor
@Observable
final class ServerEvents {
    struct Change: OptionSet, Hashable, Sendable {
        let rawValue: Int

        /// Library status, Sonarr/Radarr tracking, synced libraries: poster
        /// badges, title pages, the calendar, Discover's Recently added.
        static let library = Change(rawValue: 1 << 0)
        static let requests = Change(rawValue: 1 << 1)
        static let notifications = Change(rawValue: 1 << 2)
        static let favorites = Change(rawValue: 1 << 3)
        /// Integrations, jobs, TMDb configuration.
        static let settings = Change(rawValue: 1 << 4)
        /// Household accounts.
        static let users = Change(rawValue: 1 << 5)
        /// What browse lists are built from: the TMDb/TVDB keys, the library
        /// connections (Plex, Jellyfin, Sonarr/Radarr) and syncs. Always
        /// recorded alongside `.settings`, so a list can reload for these
        /// without also reloading for a Discord webhook or an ntfy topic.
        static let catalog = Change(rawValue: 1 << 6)

        static let all: Change = [.library, .requests, .notifications, .favorites, .settings, .users, .catalog]
    }

    enum Source: Sendable {
        /// This app changed something (a `MarqueeAPI` mutation).
        case mutation
        /// Polling noticed the server's state moved on its own.
        case server
    }

    private(set) var library = 0
    private(set) var requests = 0
    private(set) var notifications = 0
    private(set) var favorites = 0
    private(set) var settings = 0
    private(set) var users = 0
    private(set) var catalog = 0

    /// The same counters, but only advanced by `.server` changes. A screen that
    /// already applies its own action in place keys its reload off these, so
    /// acting on a title never re-renders (and re-scrolls) the page you acted on.
    private(set) var remoteLibrary = 0
    private(set) var remoteRequests = 0
    private(set) var remoteNotifications = 0
    private(set) var remoteFavorites = 0
    private(set) var remoteSettings = 0
    private(set) var remoteUsers = 0
    private(set) var remoteCatalog = 0

    /// Called after a `.mutation` is recorded, so the badge poller can
    /// refresh right away instead of waiting for its next tick.
    @ObservationIgnored var onMutation: ((Change) -> Void)?

    func record(_ change: Change, source: Source = .mutation) {
        if change.contains(.library) { library &+= 1 }
        if change.contains(.requests) { requests &+= 1 }
        if change.contains(.notifications) { notifications &+= 1 }
        if change.contains(.favorites) { favorites &+= 1 }
        if change.contains(.settings) { settings &+= 1 }
        if change.contains(.users) { users &+= 1 }
        if change.contains(.catalog) { catalog &+= 1 }
        if source == .server {
            if change.contains(.library) { remoteLibrary &+= 1 }
            if change.contains(.requests) { remoteRequests &+= 1 }
            if change.contains(.notifications) { remoteNotifications &+= 1 }
            if change.contains(.favorites) { remoteFavorites &+= 1 }
            if change.contains(.settings) { remoteSettings &+= 1 }
            if change.contains(.users) { remoteUsers &+= 1 }
            if change.contains(.catalog) { remoteCatalog &+= 1 }
        }
        if source == .mutation, !change.isEmpty {
            onMutation?(change)
        }
    }

    /// One value that moves whenever any of `areas` does, from either source.
    /// Reading it inside a view body (or `.task(id:)`) observes exactly those
    /// counters. Use this on a screen whose list *should* rebuild after the
    /// viewer's own action (Favorites losing a card, the request queue losing
    /// a row); use `remoteRevision(of:)` on a screen that updates in place.
    func revision(of areas: Change) -> Int {
        var total = 0
        if areas.contains(.library) { total &+= library }
        if areas.contains(.requests) { total &+= requests }
        if areas.contains(.notifications) { total &+= notifications }
        if areas.contains(.favorites) { total &+= favorites }
        if areas.contains(.settings) { total &+= settings }
        if areas.contains(.users) { total &+= users }
        if areas.contains(.catalog) { total &+= catalog }
        return total
    }

    /// Moves only when the server's own state changed under us (a download
    /// finished, someone requested something from the website) — never for a
    /// mutation this app made.
    func remoteRevision(of areas: Change) -> Int {
        var total = 0
        if areas.contains(.library) { total &+= remoteLibrary }
        if areas.contains(.requests) { total &+= remoteRequests }
        if areas.contains(.notifications) { total &+= remoteNotifications }
        if areas.contains(.favorites) { total &+= remoteFavorites }
        if areas.contains(.settings) { total &+= remoteSettings }
        if areas.contains(.users) { total &+= remoteUsers }
        if areas.contains(.catalog) { total &+= remoteCatalog }
        return total
    }
}

File: windows/Marquee.Core/Api/ServerEvents.cs
namespace Marquee.Core.Api;

/// <summary>
/// The areas of server state a call can change. A screen reloads for the
/// areas it shows, so approving a request refreshes the queue and the
/// library badges without also re-fetching Settings.
/// </summary>
[Flags]
public enum ServerChange
{
    None = 0,

    /// <summary>Library status, Sonarr/Radarr tracking, synced libraries: poster badges, title pages, the calendar, Discover's Recently added.</summary>
    Library = 1 << 0,

    Requests = 1 << 1,
    Notifications = 1 << 2,
    Favorites = 1 << 3,

    /// <summary>Household accounts.</summary>
    Users = 1 << 4,

    /// <summary>Connections, keys and webhooks under Settings, and what browse lists are built from (TMDb, Plex, Sonarr).</summary>
    Integrations = 1 << 5,

    Jobs = 1 << 6,

    All = Library | Requests | Notifications | Favorites | Users | Integrations | Jobs,
}

public enum ServerChangeSource
{
    /// <summary>This app changed something (a <see cref="MarqueeApi"/> mutation).</summary>
    Mutation,

    /// <summary>Polling noticed the server's state moved on its own.</summary>
    Server,
}

public sealed class ServerChangedEventArgs(ServerChange change, ServerChangeSource source) : EventArgs
{
    public ServerChange Change { get; } = change;
    public ServerChangeSource Source { get; } = source;
}

/// <summary>
/// Change counters screens key their reloads off. Every successful mutation
/// through <see cref="MarqueeApi"/> bumps the areas it touches; the badge
/// poller bumps <see cref="ServerChange.Notifications"/> and
/// <see cref="ServerChange.Requests"/> when it sees the counts move on the
/// server (a download finished, a member requested something from the
/// website).
///
/// Thread-safe: a mutation completes on whatever thread the transport
/// resumed on. <see cref="Changed"/> is raised on that thread too, so a view
/// model dispatches to the UI thread itself.
/// </summary>
public sealed class ServerEvents
{
    private static readonly ServerChange[] Areas =
    [
        ServerChange.Library,
        ServerChange.Requests,
        ServerChange.Notifications,
        ServerChange.Favorites,
        ServerChange.Users,
        ServerChange.Integrations,
        ServerChange.Jobs,
    ];

    private readonly object gate = new();
    private readonly int[] revisions = new int[Areas.Length];
    private readonly int[] remoteRevisions = new int[Areas.Length];

    /// <summary>Raised after a non-empty change is recorded, from either source.</summary>
    public event EventHandler<ServerChangedEventArgs>? Changed;

    public void Record(ServerChange change, ServerChangeSource source = ServerChangeSource.Mutation)
    {
        if (change == ServerChange.None)
        {
            return;
        }
        lock (gate)
        {
            for (var index = 0; index < Areas.Length; index++)
            {
                if (change.HasFlag(Areas[index]))
                {
                    revisions[index]++;
                    if (source == ServerChangeSource.Server)
                    {
                        remoteRevisions[index]++;
                    }
                }
            }
        }
        Changed?.Invoke(this, new ServerChangedEventArgs(change, source));
    }

    /// <summary>
    /// One value that moves whenever any of <paramref name="areas"/> does,
    /// from either source. Use it on a screen whose list should rebuild
    /// after the viewer's own action (Favorites losing a card, the request
    /// queue losing a row); use <see cref="RemoteRevision"/> on a screen
    /// that updates in place.
    /// </summary>
    public int Revision(ServerChange areas) => Sum(revisions, areas);

    /// <summary>
    /// Moves only when the server's own state changed under us, never for a
    /// mutation this app made, so acting on a title never re-renders (and
    /// re-scrolls) the page you acted on.
    /// </summary>
    public int RemoteRevision(ServerChange areas) => Sum(remoteRevisions, areas);

    private int Sum(int[] counters, ServerChange areas)
    {
        lock (gate)
        {
            var total = 0;
            for (var index = 0; index < Areas.Length; index++)
            {
                if (areas.HasFlag(Areas[index]))
                {
                    total += counters[index];
                }
            }
            return total;
        }
    }
}
